### ISSUE TITLE
RELAX - Null hypothesis AIC values

### ISSUE THREAD

**Issue by kavithamallayaa** [veg/hyphy#956]
Hi Sergei,

I am trying to find how selection intensity varies along a  few branches in a phylogeny. My test branches contains few tips, so I used two different relax analysis.
Analysis 1: 
test : branches without the tips     reference: remaining tree including the tips of the test branches
Analysis 2:
test : branches along with the tips   reference : remaining tree

I suppose the null model is same omega for all branches and so the AIC values of the null model for the both above analysis is the same as the overall dataset is the same only branches chosen as test are different. Can you please explain how this works? In case if need the input ad output files I am ready to send.

with regards
Kavitha

**Comment by spond** [comment]
Dear @kavithamallayaa,

Yes, the null models for the standard RELAX tests are the same regardless of how you partition the branches **unless** you have different sets of **unlabeled** branches. Are you getting different scores? If so, can you paste in RELAX _terminal_ output for Analyses 1 and 2?

Best,
Sergei

**Comment by kavithamallayaa** [comment]
Hi Sergei,
Yes, I am getting different scores for analysis 1 and 2. These analysis are different only for the partition of the branches as you mentioned. So it will better if have look into output files. I have the json files and also slurm output of the analysis. I am not able to attach these files nor convert them, so how can I send it to you? I am not aware of the RELAX terminal which you have mentioned. Please suggest a way to work this out.
regards
Kavitha

**Comment by kavithamallayaa** [comment]
Hi Sergei,
I have attached the zip files of the Json and slurm output. 
ntips- analysis 1( tips were not included in the test)
atips- analysis 2( tips were included in the test)


[ntips.zip](https://github.com/veg/hyphy/files/3159393/ntips.zip)
[atips.zip](https://github.com/veg/hyphy/files/3159394/atips.zip)

Thanks for the support
regards
Kavitha

**Comment by spond** [comment]
Dear @kavithamallayaa,

Thanks for sharing the files. Because a mixture model, RELAX is susceptible to convergence issues, i.e. the end result will often depend on the starting point. RELAX (minimal) will first run the **alternative** model, and then the null model, using the results from the alternative fit as a _starting_ point. Therefore for your two runs, the starting points for the null optimization were different, and the final MLEs were also somewhat different. 

They are not too dissimilar, however, so that's encouraging. 

```
Run 1 Null Log (L) = -86334.78
Run 2 Null Log (L) = -86340.65
```

Inferred rate distributions for the null

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.006     |   90.586    |                                   |
|        Negative selection         |     0.349     |    8.184    |                                   |
|      Diversifying selection       |    77.194     |    1.230    |                                   |

and

|         Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.006     |   90.324    |                                   |
|        Negative selection         |     0.348     |    8.497    |                                   |
|      Diversifying selection       |    105.382    |    1.179    |                                   |

I have put some work into improving the convergence properties of RELAX since v2.1 (which is what you have); if you download the develop branch (see build instructions at https://github.com/veg/hyphy-analyses/blob/master/README.md), you can try the current version of RELAX to see if the convergence behavior is better.

Best,
Sergei

**Comment by kavithamallayaa** [comment]
Hi Sergei,
Let me try using the current version.
Thanks for the support.

Cheers
Kavitha

**Comment by kavithamallayaa** [comment]
Hi Sergei,

I downloaded the hyphy-develop and tried it. Every time I run the analysis, just after entering the RELAX analysis type it says "obtaining branch length....." as usually but immediately followed by "Illegal instruction". I don't know how to fix this issue.  I am using the same files and procedure as the previous one, then why I encounter this problem and how to overcome this.

with regards
Kavitha

**Comment by spond** [comment]
Dear @kavithamallayaa,

This sounds like a bug in the current developer branch. Could you share the file and analysis settings with me so I can reproduce and fix the issue?

Best,
Sergei

**Comment by kavithamallayaa** [comment]
Hi Sergei,
Please find the input files[ one fasta file and 2 tree files for two different models ] in the attachment.
Analysis settings : 
Selection analysis, Relax, Universal genetic code, test : secondary ,rate :3 , Minimal

If you need any further details, pls let me know.
Regards
Kavitha

[ATP.zip](https://github.com/veg/hyphy/files/3202549/ATP.zip)

**Comment by kavithamallayaa** [comment]
Dear Sergei,

I hope the bug issue is still being fixed. If you let me know the time limit, it will be easy for me to plan my things, as I am still waiting for the improved version.

Thank you for the support.

regards
Kavitha

**Comment by spond** [comment]
Dear @kavithamallayaa,

We just released version 2.5 which should improve convergence in RELAX in general. In your particular case, it seems like the 3-rate model was simply over-parameterized, and suffered convergence issues. Dropping the # of rate classes to `2` improves congruence between the runs. However in any dataset with convergence warnings, it would be normal to expect numerical instability between runs. Think of this as an inherent limitation for smaller datasets.

Best,
Sergei

---
### ATIPS

```
hyphy relax --rates 2 --alignment /Users/sergei/Downloads/ATP/atpA_nstop  --tree /Users/sergei/Downloads/ATP/atpA_atips.tre  --test atips --models Minimal
```

...skipped...


### * Detected convergence issues (negative LRT). Refitting the alterative/null model pair from a new starting point
* Log(L) = -86215.56, AIC-c = 172983.15 (275 estimated parameters)
* Relaxation/intensification parameter (K) =     1.08
* The following rate distribution was inferred for **test** branches

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.010     |   96.892    |                                   |
|      Diversifying selection       |     2.027     |    3.108    |                                   |

* The following rate distribution was inferred for **reference** branches

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.015     |   96.892    |                                   |
|      Diversifying selection       |     1.919     |    3.108    |                                   |

Fitting the null (K := 1) model
* Log(L) = -86223.02, AIC-c = 172996.06 (274 estimated parameters)
* The following rate distribution for test/reference branches was inferred

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.014     |   96.879    |                                   |
|      Diversifying selection       |     1.925     |    3.121    |                                   |

----
## Test for relaxation (or intensification) of selection [RELAX]
Likelihood ratio test **p =   0.0001**.
>Evidence for *intensification of selection* among **test** branches _relative_ to the **reference** branches at P<=0.05
----

---
### NTIPS

* Log(L) = -86206.54, AIC-c = 172965.11 (275 estimated parameters)
* Relaxation/intensification parameter (K) =     0.67
* The following rate distribution was inferred for **test** branches

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.055     |   96.700    |                                   |
|      Diversifying selection       |     1.450     |    3.300    |                                   |

* The following rate distribution was inferred for **reference** branches

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.013     |   96.700    |                                   |
|      Diversifying selection       |     1.739     |    3.300    |                                   |


### Fitting the null (K := 1) model
* Log(L) = -86222.89, AIC-c = 172995.81 (274 estimated parameters)
* The following rate distribution for test/reference branches was inferred

|          Selection mode           |     dN/dS     |Proportion, %|               Notes               |
|-----------------------------------|---------------|-------------|-----------------------------------|
|        Negative selection         |     0.014     |   96.896    |                                   |
|      Diversifying selection       |     1.941     |    3.104    |                                   |

----
## Test for relaxation (or intensification) of selection [RELAX]
Likelihood ratio test **p =   0.0000**.
>Evidence for *relaxation of selection* among **test** branches _relative_ to the **reference** branches at P<=0.05
